libs/community/langchain_community/tools/firecrawl/tool.py

```diff
@@ -0,0 +1,127 @@
+from pydantic import BaseModel, Field
+
+# Configure logging
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+
+
+class ScrapeWebsiteInput(BaseModel):
+    base_url: str = Field(..., description="The base URL to scrape.")
+
+
+class FirecrawlScrapeWebsiteTool(BaseTool):
+    name: str = "scrape_website"
+    description: str = (
+        "A tool for scraping websites and extracting markdown content. "
+        "Provide the URL of the website to scrape and indicate if the scraping rate "
+        "should be limited. The parameters required are:\n"
+        "- 'base_url': The base URL of the website to scrape.\n"
+        "- 'limit_rate': A boolean specifying whether to limit the scraping rate."
+    )
+    args_schema: Type[BaseModel] = ScrapeWebsiteInput
+    return_direct: bool = True
+
+    api_key: str = Field(
+        default_factory=lambda: os.getenv("FIRECRAWL_API_KEY", ""),
+        description="API key for Firecrawl",
+    )
+    limit_rate: bool = Field(
+        default_factory=lambda: os.getenv("FIRECRAWL_LIMIT_RATE", "True")
+        .strip()
+        .lower()
+        == "true",
+        description="Limit scraping rate",
+    )
+
+    def __init__(
+        self, api_key: Optional[str] = None, limit_rate: Optional[bool] = None
+    ):
+        super().__init__()
+        self.api_key = api_key or self.api_key
+        self.limit_rate = limit_rate if limit_rate is not None else self.limit_rate
+        if not self.api_key:
+            raise ValueError(
+                "API key for Firecrawl is required. Please set the "
+                "FIRECRAWL_API_KEY environment variable or pass it to the "
+                "constructor."
+            )
+
+    def _run(
+        self,
+        base_url: str,
+        run_manager: Optional[CallbackManagerForToolRun] = None,
+    ) -> str:
+        """Use the tool synchronously."""
+        return asyncio.run(self.scrape_all_urls(base_url))
+
+    async def _arun(
+        self,
+        base_url: str,
+        run_manager: Optional[AsyncCallbackManagerForToolRun] = None,
+    ) -> str:
+        """Use the tool asynchronously."""
+        return await self.scrape_all_urls(base_url)
+
+    async def scrape_all_urls(self, base_url: str) -> str:
+        app = self.get_firecrawl_app()
+
+        urls = self.map_website(app, base_url)
+        if not urls:
+            return "No URLs found. Please check if the base URL is correct."
+
+        markdown_content = ""
+        for i, url in enumerate(urls):
+            content = await self.async_scrape_url(app, url)
+            markdown_content += f"# {url}\n\n{content}\n\n---\n\n"
+
+            # Flexible rate limiting
+            if self.limit_rate and (i + 1) % 10 == 0:
+                logging.info("Rate limiting: Sleeping for 60 seconds")
+                await asyncio.sleep(60)
+
+        return markdown_content
+
+    def get_firecrawl_app(self) -> FirecrawlApp:
```

Comment:
this should be done once on init and stored in `self.app` or the like

Reply:
done

Reply:
ah, we should just remove this altogether then! and use self.app anywhere we need the app